servlet-engine/src/main/java/io/micronaut/servlet/engine/body/InputStreamByteBody.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    static InputStreamByteBody create(ExtendedInputStream stream, OptionalLong length, Executor ioExecutor) {
+        return new InputStreamByteBody(new Context(length, ioExecutor), stream);
+    }
+
+    @Override
+    public @NonNull CloseableByteBody allowDiscard() {
+        stream.allowDiscard();
+        return this;
+    }
+
+    @Override
+    public void close() {
+        if (stream != null) {
+            stream.close();
+            stream = null;
+        }
+    }
+
+    @Override
+    public @NonNull CloseableByteBody split(SplitBackpressureMode backpressureMode) {
+        if (stream == null) {
+            failClaim();
+        }
+        StreamPair.Pair pair = StreamPair.createStreamPair(stream, backpressureMode);
+        stream = pair.left();
+        return new InputStreamByteBody(context, pair.right());
+    }
+
+    @Override
+    public @NonNull OptionalLong expectedLength() {
+        return context.expectedLength();
+    }
+
+    @Override
+    public @NonNull ExtendedInputStream toInputStream() {
+        ExtendedInputStream s = stream;
+        if (s == null) {
+            failClaim();
+        }
+        stream = null;
+        return s;
+    }
+
+    @Override
+    public @NonNull Flux<byte[]> toByteArrayPublisher() {
+        ExtendedInputStream s = toInputStream();
+        Sinks.Many<byte[]> sink = Sinks.many().unicast().onBackpressureBuffer();
+        return sink.asFlux()
+            .doOnRequest(req -> {
+                long remaining = req;
+                while (remaining > 0) {
+                    @Nullable byte[] arr;
+                    try {
+                        arr = s.readSome();
+                    } catch (IOException e) {
+                        sink.tryEmitError(e);
+                        break;
+                    }
+                    if (arr == null) {
+                        sink.tryEmitComplete();
+                        break;
+                    } else {
+                        remaining--;
+                        sink.tryEmitNext(arr);
+                    }
+                }
+            })
+            .doOnTerminate(s::close)
+            .doOnCancel(s::close)
+            .subscribeOn(Schedulers.fromExecutor(context.ioExecutor()));
+    }
+
+    @Override
+    public @NonNull Publisher<ByteBuffer<?>> toByteBufferPublisher() {
+        return toByteArrayPublisher().map(ByteArrayByteBuffer::new);
+    }
+
+    @Override
+    public CompletableFuture<? extends CloseableAvailableByteBody> buffer() {
+        ExtendedInputStream s = toInputStream();
+        return CompletableFuture.supplyAsync(() -> {
+            try (ExtendedInputStream t = s) {
+                return new AvailableByteArrayBody(t.readAllBytes());
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }, context.ioExecutor);
+    }
+
+    private record Context(
+        OptionalLong expectedLength,
```

Comment:
probably should just be `@Nullable Long` not sure there is much value in this wrapper

Reply:
It's just easier with OptionalLong since that's what the expectedLength API returns.